src/middlewared/middlewared/plugins/ports/ports.py

```diff
@@ -0,0 +1,36 @@
+from middlewared.service import Service, ValidationErrors
+
+
+class PortService(Service):
+
+    DELEGATES = {}
+    SYSTEM_USED_PORTS = [
+        {'title': 'System', 'ports': [67, 123, 3702, 5353, 6000], 'namespace': 'system'},
+    ]
+
+    class Config:
+        private = True
+
+    async def register_attachment_delegate(self, delegate):
+        self.DELEGATES[delegate.namespace] = delegate
+
+    async def get_in_use(self):
+        ports = self.SYSTEM_USED_PORTS.copy()
```

Comment:
There is no need for a shallow copy either. Here's what I was thinking about doing

```
ports = []
for delegate in self.DELEGATES.values():
    if used_ports := await delegate.get_ports():
        ports.append({
            'namespace': delegate.namespace,
            'title': delegate.title,
            'ports': used_ports,
        })

ports += self.SYSTEM_USED_PORTS
return ports